core/privc/gc_bit.cc

```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "gc_bit.h"
+
+namespace privc {
+
+void garbled_and(const TensorBlock* a, const TensorBlock* b, TensorBlock* ret) {
+    auto block_shape = ret->shape();
+    auto shape = block_shape;
+    shape.erase(shape.begin());
+    auto j0 = tensor_factory()->template create<int64_t>(shape);
+    auto j1 = tensor_factory()->template create<int64_t>(shape);
+    ot()->garbled_and_ctr(j0.get());
+    ot()->garbled_and_ctr(j1.get());
+
+    auto j0_ = tensor_factory()->template create<int64_t>(block_shape);
+    auto j1_ = tensor_factory()->template create<int64_t>(block_shape);
+    common::to_block(j0.get(), j0_.get());
+    common::to_block(j1.get(), j1_.get());
+
+    auto pa = tensor_factory()->template create<u8>(shape);
+    auto pb = tensor_factory()->template create<u8>(shape);
+    block_lsb(a, pa.get());
+    block_lsb(b, pb.get());
+
+    if (party() == 0) {
+        auto garbled_delta = tensor_factory()->template create<int64_t>(block_shape);
+        ot()->garbled_delta(garbled_delta.get());
+        auto mask_a = tensor_factory()->template create<int64_t>(block_shape);
+        a->bitwise_xor(garbled_delta.get(), mask_a.get());
+        std::pair<const TensorBlock*, const TensorBlock*> a_pair(a, mask_a.get());
+        std::pair<TensorBlock*, TensorBlock*> j0_pair(j0_.get(), j0_.get());
+        auto t_first = tensor_factory()->template create<int64_t>(block_shape);
```

Comment:
what does t, tg, wg, te, we stand for?
And please add some comment to show what each code segmentation is doing.

Reply:
maybe comment on the code that naming come from the paper

Reply:
agree with Kai, and a better idea is to comment the algorithm sketch containing the variables